### ISSUE TITLE
[Bug] bin tools are using jstrencode(1) when they should be decoding JSON encoded strings

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/temp-test-ioccc#2752]
### Is there an existing issue for this?

- [X] I have searched for existing issues and did not find anything like this

### Describe the bug

Tools such as:

- bin/cvt-submission.sh
- bin/entry2csv.sh
- bin/gen-authors.sh
- bin/gen-location.sh
- bin/gen-years.sh
- bin/output-index-author.sh
- bin/output-year-index.sh
- bin/subst.entry-index.sh

Are using `jstrencode(1)` when they need to take JSON encoded strings and decode them into "real strings" for use in markdown and HTML pages.

### What you expect

Tools such as:

- bin/cvt-submission.sh
- bin/entry2csv.sh
- bin/gen-authors.sh
- bin/gen-location.sh
- bin/gen-years.sh
- bin/output-index-author.sh
- bin/output-year-index.sh
- bin/subst.entry-index.sh

would use `jstrdecode(1)` to decode JSON encoded strings, converting them into "real" strings.

### Environment

- OS: n/a
- Device: n/a
- Compiler: n/a

### Anything else?

Things may have gone amiss with commit e58bd97211be82e122da55ba19023f5ba61a6843 (dated: Fri Nov 1 06:45:02 2024 -0700)

This seems to be linked to commit 0a7c9673fa7797f1e9c2c87dea377edb03816f03 (dated: Thu Oct 31 11:57:38 2024 -0700) from the "other repo".

## UPDATE 0

This is a **Great Fork Merge** show stopper.

**Comment by lcn2** [comment]
We remain concerned with the issue of JSON string encoding and decoding.

We suspect that the details of [GH-issuecomment-2471493104](https://github.com/xexyl/jparse/issues/14#issuecomment-2471493104) were either unclear or glossed over.

We don't have access to the shell now, but when we return we plan to do more testing.   Nevertheless what we saw in the code concerned us enough to file this bug and halt the **Great Fork Merge** process.

**Comment by xexyl** [comment]
Okay. Without having the time to read this I am admittedly confused because taking UTF-8 to Unicode is encoding and it seemed to work. 

But I will have to look at this tomorrow. Sorry.

**Comment by xexyl** [comment]
I just read it and it doesn't make sense. It sounds like a terminology issue but taking a UTF-8 code point and converting it into a Unicode symbol is encoding not decoding. 

Please explain what you are getting at. I will look at this tomorrow.

**Comment by lcn2** [comment]
UTF-8 is an encoding of Unicode stuff. 

JSON string encoding of "real" strings, decoding of JSON encoded strings into "real" strings  is another matter.

When `jstrdecode(1)` decodes a JSON encoded string such as:

```json
"\uD83D\uDD25"
```

one expects to get the "real" string of:

```
🔥
```

### P.S.

The man page for `jstrencode(1)`, as we think we mentioned before, seems wrong.  It reads:

```
jstrencode encodes JSON decoded strings given on the command line.
```

The term "JSON decoded strings" is somewhat meaningless.  JSON doesn't decode strings.  The so-called JSON specification requires all strings to be encoded, and thus produce JSON encoded strings.  See  [GH-ssuecomment-2471493104](https://github.com/xexyl/jparse/issues/14#issuecomment-2471493104).

The man page for `jstrencode(1)` should read something like:

```
jstrencode encodes strings into JSON encoded strings accoding to the so-called JSON specification.
```

**Comment by xexyl** [comment]
All sources say the opposite of what you're saying though. That's why I swapped the terms. 

The man page can be updated of course as can documentation.

**Comment by lcn2** [comment]
> I just read it and it doesn't make sense. It sounds like a terminology issue but taking a UTF-8 code point and converting it into a Unicode symbol is encoding not decoding.
> 
> Please explain what you are getting at. I will look at this tomorrow.

JSON requires strings to be encoded.  This JSON string encoding has nil to do with how UTF-8 encodes Unicode stuff.  

Quoting from [GH-issuecomment-2471493104](https://github.com/xexyl/jparse/issues/14#issuecomment-2471493104):

JSON string encoding, at a minimum, requires the string to be surrounded by double quotes. At a minimum, encoding will result in the prepending and appending a double quote character.

JSON string encoding ALSO requires one to convert things like ASCII newlines into "\n".  There are other important back-slashing requirements such as dealing with double quotes are within the "real" string, backslashes, tabs, etc.  (that need to be handed during the JSON string encoding process)

Quoting from [GH-issuecomment-2471493104](https://github.com/xexyl/jparse/issues/14#issuecomment-2471493104) again:

Encoding this "real" string:

```
This "string" has a newline
in the middle and at the end
```

into this JSON encoded string:

```json
"This \"string\" has a newline\nin the middle and at the end\n"
```

Now the question of what to do with non-ASCII stuff **IS** a matter for Unicode / UTF-8 encoding and decoding.    But that is **NOT** JSON string encoding, nor is it decoding of JSON encoded strings.

The [issue 13 in the other repo](https://github.com/xexyl/jparse/issues/13) raised a concern about JSON encoded strings that contained \uHexHexHexHex-stuff was being handed as a side effect of decoding JSON encoded strings.

Tools such as `jsp(1)` formatted or "pretty printed" JSON encoded strings like this:

```sh
$ echo '"œßåé"' | jsp --no-color --indent 4
"\u0153\u00df\u00e5\u00e9"
```

NOTE: The pipe input above is a JSON encoded string.  The output produced by the `jsp(1)` tool is also a JSON encoded string.   The `jsp(1)` tool happens to choose to formatted or "pretty printed" by printing \uHexHexHexHex stuff.  While it is a questionable practice from the point of view of formatting or "pretty printing", it is technically correct as BOTH JSON encoded strings are valid and are equivalent.

BTW: we think the `jsp(1)` tool makes things ugly when to does that sort of formatting.

The `jstrdecode(1)` tool and the internal JSON encoded string decoding functionality needs to be able to that the JSON encoded string:

```json
"\u0153\u00df\u00e5\u00e9"
```

and produce the original "real" string: `œßåé`.

Instead `jstrencode(1) does this!

We don't mind if, in the process of JSON string encoding (what `jstrencode(1)` should do), Unicode stuff is left along and/or converted in such a way that when one decodes (what `jstrdecode(1)` should do), produces nice looking Unicode stuff.

So by all means, let `jstrencode(1)` convert the string: 🔥by JSON encoding into "🔥" as that is valid JSON.

And by all means, let `jstrdecode(1)` convert the JSON encoded string "🔥" into the real string: 🔥

We just need to ALSO be sure that when `jstrdecode(1)` is given the JSON encoded string "\uD83D\uDD25" we get the real string: 🔥 as well.

Both "\uD83D\uDD25" and "🔥" are valid JSON encoded strings.   They should convert into the same real string 🔥 as well.

We hope this helps.

**Comment by lcn2** [comment]
> All sources say the opposite of what you're saying though. That's why I swapped the terms.

If there is a comment bug in the source, that needs to be fixed.

We looked at the comment for the `json_encode(char const *ptr, size_t len, size_t *retlen)` function in `jparse/json_parse.c` in the "other repo".  It looks correct.  Are we missing something?

**Comment by xexyl** [comment]
> > All sources say the opposite of what you're saying though. That's why I swapped the terms.
> 
> If there is a comment bug in the source comments, then those should be fixed.'

It could be a comment bug yes: when I tried to (later) correct the comments by swapping from one function to the other, it might have been a mistake. However ... I see something really odd.

> We looked at the comment for the `json_encode(char const *ptr, size_t len, size_t *retlen)` function in `jparse/json_parse.c` in the "other repo". It looks correct.

Okay so then it maybe isn't a comment bug?

> 
> Are we missing something? There certainly could have been some copy and paste errors when code for one side of encoding was converted into decoding, for example.
> 

It could be. But if the encoding comment looks correct and I swapped it then it seems like it might be right and this is not an issue? But even so see below.

> ## UPDATE 0
> We recommend to prioritize on [GH-issuecomment-2472008592](https://github.com/ioccc-src/temp-test-ioccc/issues/2752#issuecomment-2472008592) and the core of this issue first, before worrying about source code comments and man pages.

Of course.

I'll post a new comment with something funny though.

**Comment by xexyl** [comment]
Here's something funny.

All sources say that converting `'\u0f0f'` to its unicode symbol `༏` is ENCODING. But what about JavaScript's JSON object (or whatever it is)? Check this javascript out:

```javascript
const json_to_encode = {
          name: "\u0f0f"
        };
            const json_to_decode = '{"name": "\u0f0f"}';

            const json_decoded = JSON.parse(json_to_decode);
            const json_encoded = JSON.stringify(json_to_encode);

        document.writeln(json_decoded.name);
        document.write(json_encoded);
```

shows in the html file:

```
༏ {"name":"༏"}
```

which suggests that both encoding and decoding convert the `\uxxxx` to its unicode symbol!

Now given that the function we have is `utf8encode()` and given many other sources I would say that the right term to use for what we have is ENCODE and given that it does print out the correct string in the website then this is probably okay.

What do you think?

**Comment by xexyl** [comment]
> > I just read it and it doesn't make sense. It sounds like a terminology issue but taking a UTF-8 code point and converting it into a Unicode symbol is encoding not decoding.
> > Please explain what you are getting at. I will look at this tomorrow.
> 
> JSON requires strings to be encoded. This JSON string encoding has nil to do with how UTF-8 encodes Unicode stuff.
> 

Yes. But the point is that when UTF-8 `'\u0f0f'` (for example) is encoded it turns into `༏` and that's exactly what jstrencode does.

> Quoting from [GH-issuecomment-2471493104](https://github.com/xexyl/jparse/issues/14#issuecomment-2471493104):
> 
> JSON string encoding, at a minimum, requires the string to be surrounded by double quotes. At a minimum, encoding will result in the prepending and appending a double quote character.

Aha. Now I wonder about this. Perhaps the problem is that some of the options need to be moved from one tool to the other? That seems likely. Though in that case will the output of the UTF-8 code points for the website then show the right string? If it adds `"`s then it would be the wrong output, right? Perhaps that's why you want the decode tool to also do this? 


> 
> JSON string encoding ALSO requires one to convert things like ASCII newlines into "\n". There are other important back-slashing requirements such as dealing with double quotes are within the "real" string, backslashes, tabs, etc. (that need to be handed during the JSON string encoding process)
> 

Okay so it seems likely that some of the options/functionality has to be moved over too, and the names alone cannot be swapped? For instance:

```sh
$ cat nl


$ jstrdecode < nl
\n\n
$ jstrencode < nl
Warning: json_encode: found non-\-escaped char: 0x0a
Warning: jstrencode_stream: error while encoding stdin buffer
Warning: main: error while encoding processing stdin

```

?


Unless there is also some confusion with the terms?

> Quoting from [GH-issuecomment-2471493104](https://github.com/xexyl/jparse/issues/14#issuecomment-2471493104) again:
> 
> Encoding this "real" string:
> 
> ```
> This "string" has a newline
> in the middle and at the end
> ```
> 
> into this JSON encoded string:
> 
> ```json
> "This \"string\" has a newline\nin the middle and at the end\n"
> ```
> 
> The above is **JSON string encoding**.


Okay and I see...

```sh
$ jstrdecode  < foo.txt 
This \"string\" has a newline\nin the middle and at the end\n
```

But the thing is how do we determine when to encode and when to decode, then? I can see what you mean here: JSON with a `"` inside a quote is wrong unless it's escaped. But on the other hand encoding does seem to also convert a code point into its unicode symbol. This is a mess!

> 
> Now the question of what to do with non-ASCII stuff **IS** a matter for Unicode / UTF-8 encoding and decoding. But that is **NOT** JSON string encoding, nor is it decoding of JSON encoded strings.
> 

Hmm ... okay so that might be something to consider too. The question we have I think then: what do we do here? Perhaps the tool names should be reverted again but then we have to decide how to proceed with the code points?

> The [issue 13 in the other repo](https://github.com/xexyl/jparse/issues/13) raised a concern about JSON encoded strings that contained \uHexHexHexHex-stuff was being handed as a side effect of decoding JSON encoded strings.

Yes.

> 
> Tools such as `jsp(1)` formatted or "pretty printed" JSON encoded strings like this:
> 
> ```shell
> $ echo '"œßåé"' | jsp --no-color --indent 4
> "\u0153\u00df\u00e5\u00e9"
> ```
> 
> NOTE: The pipe input above is a JSON encoded string. The output produced by the `jsp(1)` tool is also a JSON encoded string. The `jsp(1)` tool happens to choose to formatted or "pretty printed" by printing \uHexHexHexHex stuff. While it is a questionable practice from the point of view of formatting or "pretty printing", it is technically correct as BOTH JSON encoded strings are valid and are equivalent.

Right. It is encoded. Which suggests that the decoded is the `\uxxxx` string. Though as my example javascript above shows it seems that it's both. Now as you say this is not json encoding/decoding so perhaps I do need to swap the names again. But in this case we do also need to determine how to encode the code points in the case of it happening. Unless we do not want this feature?




> 
> BTW: we think the `jsp(1)` tool makes things ugly when to does that sort of formatting.
> 

When it converts it to `\uxxxx` you mean?

> The `jstrdecode(1)` tool and the internal JSON encoded string decoding functionality needs to be able to that the JSON encoded string:
> 
> ```json
> "\u0153\u00df\u00e5\u00e9"
> ```
> 
> and produce the original "real" string: `œßåé`.
> 
> Instead `jstrencode(1) does this!

That's because all the sources say that that IS encoding, not decoding. Except for the example I gave above which suggests both do it.

And this is why the tools here use the jstrencode tool, not the jstrdecode tool. So minus the fact that the javascript example suggests that both encoding and decoding should print out the fire based on the emoji I gave (above), it should be good, unless you want to quibble about terminology. But since below you talk about how both should do it then it seems like that matters less if at all.

Now as you say though, the json encoding/decoding is not the same thing as unicode. On the other hand you did raise the problem of it not doing it at all.

> 
> We don't mind if, in the process of JSON string encoding (what `jstrencode(1)` should do), Unicode stuff is left alone and/or converted in such a way that when one decodes (what `jstrdecode(1)` should do), produces nice looking Unicode stuff.
> 

I wonder if there is a way to do it for both like the javascript above shows.



> So by all means, let `jstrencode(1)` convert the string: 🔥by JSON encoding into "🔥" as that is valid JSON.
> 

It does this already indeed.

> And by all means, let `jstrdecode(1)` convert the JSON encoded string "🔥" into the real string: 🔥

Well jstrdecode will take the fire emoji and output the fire emoji:

```sh
$ cat fire.json | jstrdecode -n
🔥\n
```

but it appears that the `-n` option does not work. Hmm.... I wonder why.

Still it should I believe also output the emoji for both encoding and decoding.


> 
> We just need to ALSO be sure that when `jstrdecode(1)` is given the JSON encoded string "\uD83D\uDD25" we get the real string: 🔥 as well.
> 

Is this the real problem then? The tool names should be swapped back and both should print out the fire emoji from the fire code point? This way the json encoded strings will be correctly encoded but still print out the encoded form? I think that sounds reasonable but how to go about it I'm not sure yet.


> Both "\uD83D\uDD25" and "🔥" are valid JSON encoded strings. They should convert into the same real string 🔥 as well.
> 

That's true. And that's what happens with `jstrencode(1)`. That's why the confusion. But as above... I'm not sure.

**Comment by xexyl** [comment]
I guess the following. Please correct me if I'm wrong.

0. The tools should have their named swapped again. This is because the point was for **JSON** encoding/decoding, not encoding of the strings themselves.
1. Then both decoding and encoding of the code points should turn it into the proper unicode symbol.
2. After this the documentation can be updated for both.

I have a thought on what might allow this to happen but I am not sure. I have to look at the code too.

**Comment by xexyl** [comment]
.. first step is to swap the names again. That'll be fun.

**Comment by xexyl** [comment]
Okay the filenames and terms are swapped. The next step would be to make sure that both encode and decode convert code points to unicode symbols. I think I have an idea how to do this. But I will have to go afk very soon.

**Comment by xexyl** [comment]
Made a commit ... not pushing yet. Have to go afk. Once I'm back I'll work on the unicode problem. Then we can figure out which tool belongs in the website. Hoping I can manage this today.

**Comment by xexyl** [comment]
Ugh. The real problem is that the json_encode() function (after name change) uses the table ... not the parsing manually. I don't know how to fix that yet. I'll ponder it as I'm afk (or part of the time).

**Comment by xexyl** [comment]
Well I have an idea but unfortunately it might have to be done first .. but that means the table access will be wrong. This is because the '\` for example in `\u` will be changed to `\\u`. So something has to be figured out. But it might end up that this table can be worked out to be not needed or more useful. We shall see!

**Comment by xexyl** [comment]
Just pushed the changes noted above. In a bit I will look at seeing if I can figure out the encoding/decoding of code points.

**Comment by xexyl** [comment]
Hmm .. my initial idea will not work. This is turning into a nightmare.

**Comment by xexyl** [comment]
Have another thought. Looking into it.

**Comment by xexyl** [comment]
I think i got it! Have to do more testing ...

**Comment by xexyl** [comment]
The one problem is that doing...

`sh
$ jstrencode '\'
```

fails when it should print `\\`. It seems like the encode function might need to go through the string itself, somehow, like the decoding process, but with slightly different rules.

**Comment by xexyl** [comment]
I discovered another bug too ... check this:


```sh
$ jstrencode '\a'
'\\a
```

Should not have the first character there, but just `\\a`.

**Comment by xexyl** [comment]
Taking a break from it for the moment. 

I feel like I'm getting there but I am not quite there. 

I know what is happening but I don't yet know why it's happening. 

Of course right now at least the decode version can convert to Unicode symbols but it's not correct it seems to not have it working with encoding too. 

Unsure if I will return to this today. Tomorrow is a much busier day however so if I don't get to it today I won't have as much time to work on it until Friday.

**Comment by lcn2** [comment]
> I guess the following. Please correct me if I'm wrong.
> 
> 0. The tools should have their named swapped again. This is because the point was for **JSON** encoding/decoding, not encoding of the strings themselves.
> 1. Then both decoding and encoding of the code points should turn it into the proper unicode symbol.
> 2. After this the documentation can be updated for both.
> 
> I have a thought on what might allow this to happen but I am not sure. I have to look at the code too.

At a quick glance, yes.

**Comment by xexyl** [comment]
> > I guess the following. Please correct me if I'm wrong.
> > 
> > 0. The tools should have their named swapped again. This is because the point was for **JSON** encoding/decoding, not encoding of the strings themselves.
> > 1. Then both decoding and encoding of the code points should turn it into the proper unicode symbol.
> > 2. After this the documentation can be updated for both.
> > 
> > I have a thought on what might allow this to happen but I am not sure. I have to look at the code too.
> 
> At a quick glance, yes.

Thanks. Made progress but not there yet. However the jstrdecode tool should now do what jstrencode once did though of course version changed.

Might be good to not do anything with it just yet though.

I'm going to do other things. I somehow doubt I'll return to this today but we'll see.

**Comment by lcn2** [comment]
> I discovered another bug too ... check this:
> 
> ```shell
> $ jstrencode '\a'
> '\\a
> ```
> 
> Should not have the first character there, but just `\\a`.
> 
> ## UPDATE 0
> Or perhaps that is a display issue?

We don't know which `jstrencode(1)` this refers too.  :-)

We would expect the "true" (after this bug if fixed) `jstrencode(1)` to do this:

```sh
$ echo -n '\a' | jstrencode
"\u0007"
```

**Comment by xexyl** [comment]
> > I discovered another bug too ... check this:
> > ```shell
> > $ jstrencode '\a'
> > '\\a
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should not have the first character there, but just `\\a`.
> > ## UPDATE 0
> > Or perhaps that is a display issue?
> 
> We don't know which `jstrencode(1)` this refers too. :-)
> 
> We would expect the "true" (after this bug if fixed) `jstrencode(1)` to do this:
> 
> ```shell
> $ echo -n '\a' | jstrencode
> "\u0007"
> ```

Okay well that's far more problematic I think than say `\a`. Well .. maybe not. Perhaps the table has it, now I think on it. I think I might have to swap the order of the way things are done but that'll not happen today. Perhaps not much work on it tomorrow but Friday I should have more time.

**Comment by xexyl** [comment]
As far as what version I was referring to: it is after swapping the names back to reflect it's **JSON** encoding.

**Comment by xexyl** [comment]
Actually .. the jstrencode and printf of `\a` already works though it has the ' there.

**Comment by xexyl** [comment]
Okay so .. and this is the last thing for now I think.

What is committed (earlier today) works for jstrencode/jstrdecode. With one exception (I believe one). The encoding does not translate code points to a unicode symbol. It is jstrdecode that does that now but it seems that both are supposed to do it.

I think the current changes I have are actually useless or almost useless. I will have to ponder it before I work on it again .. or at least take a break before I do.

But at least jstrdecode should do what the website needs and jstrencode does what you expect above too. It is getting the encoding of code points that will be a problem I think.

**Comment by xexyl** [comment]
Oddly that command line does not show the correct output. I could have sworn it did though. Hmm .... I wonder why. I'll have to investigate this. I have a bit of time but in about 2.25 hours I have to be afk a while. Then later in the afternoon I visit with the Ents again (well actually the audiologist) but the good news about that is at least I will get my other hearing aid back. Bad news is there is usually a long wait in the waiting room and it's small and crowded. Let's just say it's like a small ent-house without the Ents :(

**Comment by xexyl** [comment]
In fact it seems like that might not have happened before the swap ... hm. I wonder.

**Comment by xexyl** [comment]
Actually going back even before the initial swap it still doesn't work :(

So it seems like there might be another bug?

**Comment by xexyl** [comment]
Oh! I know why. It's the echo; using `printf(1)` and it works fine.

**Comment by xexyl** [comment]
Great .. head detached. Might have to clone repo again on server.

**Comment by xexyl** [comment]
Here's a bug though that I just uncovered. Seems old at a guess:

```sh
$ printf '\b' | ./jstrencode 
\b
```

:(

Well have to go afk a bit now ... back later I hope.

**Comment by xexyl** [comment]
Might (not sure) have made progress .. have to go for a bit though.

**Comment by xexyl** [comment]
Okay well one thought is this.

Since it is not STRICTLY a UTF-8 encoder but rather making a JSON encoder, it could be considered 'okay', maybe, to not have it encoded the code points. But on the other hand other encoders show both. The question is whether those also show `\\` for `\`. I have to test this out.

**Comment by xexyl** [comment]
Okay this is extremely interesting.

```javascript
const json_to_encode = {
          name: "\u0f0f\\"
        };
            const json_to_decode = '{"name": "\u0f0f"}';

            const json_decoded = JSON.parse(json_to_decode);
            const json_encoded = JSON.stringify(json_to_encode);

        document.writeln(json_decoded.name);
        document.write(json_encoded);

```

This shows just:

```
༏ {"name":"༏\\"}
```

So the question is how to even deal with the `\`. I mean the `\` escapes the next character, right?

I guess it might be different if the `\` is in the middle of the string? Let's see...

Nope. And I guess this is because it turns it into an escape character which might or might not be valid.

So am I inputting things wrong or is the concept of the `jstrencode(1)` tool wrong? This has to be answered before proceeding (possibly).

**Comment by xexyl** [comment]
We seem to have a much bigger problem. See the bug report I just opened which has at least three problems.

https://github.com/xexyl/jparse/issues/28.

Now since the website relies on the restored jstrdecode and not jstrencode this might not be a problem for the website which would be good but the issues I have uncovered seem to be really bad.

Please let me know what you want me to do. My guess is I should sync the current code to mkiocccentry and then also test `make www` in the website (after updating the tools to use the proper version of `jstrdecode(1)`.

I can do the testing later today possibly but if not today then tomorrow.

Nonetheless the jstrencode tool seems to be broken in numerous ways. Any help would be **EXTREMELY** appreciated.

**Comment by lcn2** [comment]
> ## QUESTION THAT HAS TO BE ANSWERED
> Okay this is extremely interesting.
> 
> ```js
> const json_to_encode = {
>           name: "\u0f0f\\"
>         };
>             const json_to_decode = '{"name": "\u0f0f"}';
> 
>             const json_decoded = JSON.parse(json_to_decode);
>             const json_encoded = JSON.stringify(json_to_encode);
> 
>         document.writeln(json_decoded.name);
>         document.write(json_encoded);
> ```
> 
> This shows just:
> 
> ```
> ༏ {"name":"༏\\"}
> ```
> 
> So the question is how to even deal with the `\`. I mean the `\` escapes the next character, right?
> 
> I guess it might be different if the `\` is in the middle of the string? Let's see...
> 
> Nope. And I guess this is because it turns it into an escape character which might or might not be valid.
> 
> So am I inputting things wrong or is the concept of the `jstrencode(1)` tool wrong? This has to be answered before proceeding (possibly).

We are not sure we see the issue here.  This is valid JSON:

```json
{"name":"༏\\"}
```

The so-called JSON spec requires a backslash to be JSON encoded into a double backslash.

To JSON encode this string, `༏\` one gets: "༏\\"

To decode the JSON encoded string: "༏\\" one should get: `༏\` back.

**Comment by xexyl** [comment]
> > ## QUESTION THAT HAS TO BE ANSWERED
> > Okay this is extremely interesting.
> > ```js
> > const json_to_encode = {
> >           name: "\u0f0f\\"
> >         };
> >             const json_to_decode = '{"name": "\u0f0f"}';
> > 
> >             const json_decoded = JSON.parse(json_to_decode);
> >             const json_encoded = JSON.stringify(json_to_encode);
> > 
> >         document.writeln(json_decoded.name);
> >         document.write(json_encoded);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This shows just:
> > ```
> > ༏ {"name":"༏\\"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So the question is how to even deal with the `\`. I mean the `\` escapes the next character, right?
> > I guess it might be different if the `\` is in the middle of the string? Let's see...
> > Nope. And I guess this is because it turns it into an escape character which might or might not be valid.
> > So am I inputting things wrong or is the concept of the `jstrencode(1)` tool wrong? This has to be answered before proceeding (possibly).
> 
> We are not sure we see the issue here. This is valid JSON:
> 
> ```json
> {"name":"༏\\"}
> ```
> 
> The so-called JSON spec requires a backslash to be JSON encoded into a double backslash.
> 
> To JSON encode this string, `༏\` one gets: "༏\"
> 
> To decode the JSON encoded string: "༏\" one should get: `༏\` back.

Yes but JavaScript decoding of json does not have that. I don't understand it. Well it might be for if you want a literal backslash. However I think the issue was something else, something that is a much bigger problem (related to the escaping of characters in the bug report I opened today) and will require other changes. Thus I think this right here is not as important to consider.

Thankfully jstrdecode does work at least and that's what the website needs.

**Comment by lcn2** [comment]
> Thankfully jstrdecode does work at least and that's what the website needs

The tools are still broken as noted in comments in the other repo.

**Comment by xexyl** [comment]
> > We seem to have a much bigger problem. See the bug report I just opened which has at least three problems.
> > [xexyl/jparse#28](https://github.com/xexyl/jparse/issues/28).
> > Now since the website relies on the restored jstrdecode and not jstrencode this might not be a problem for the website which would be good but the issues I have uncovered seem to be really bad.
> > Please let me know what you want me to do. My guess is I should sync the current code to mkiocccentry and then also test `make www` in the website (after updating the tools to use the proper version of `jstrdecode(1)`.
> > I can do the testing later today possibly but if not today then tomorrow.
> > Nonetheless the jstrencode tool seems to be broken in numerous ways. Any help would be **EXTREMELY** appreciated.
> 
> Should be revert the code base to before these changes occurred?
> 
> I.e., to revert the "other repo" back to: commit 65f3d4ceecce4f9307ec82cd0ec0cb8e5a39b003
> 
> and to never this repo back to: commit [184cc9b](https://github.com/ioccc-src/temp-test-ioccc/commit/184cc9be9a8cc7ec2f878aa611d78d92b188124e)
> 
> We release that there are a few typo / fixes that are not related to the jstrencode / jstrdecode tools after those commits that might have to be reapplied. But doing this would avoid and restore the web site to a state that worked well. Some very selective fixes could be applied back on top of the revert and avoid the problematic code.
> 
> We did a test of this reversion and besides an issue of `2020/kurdyukov4` in `years.html` the can be cleaned up, the site works.

That example you give should work with jstrdecode as it stands.

**Comment by xexyl** [comment]
> > Thankfully jstrdecode does work at least and that's what the website needs
> 
> The tools are still broken as noted in comments in the other repo.

See my reply there. It's not as simple as that. In some cases quotes do need to be escaped and in other cases they do not.

So the decoder assumes that it's inside a string when your input was the string itself.

**Comment by xexyl** [comment]
... but what I meant was the tool with the purpose of utf-8 to unicode works.

**Comment by xexyl** [comment]
I think it would be a big mistake to revert. Many other fixes were made, some quite important.

The jstrdecode works fine for utf-8 to unicode.

**Comment by xexyl** [comment]
So please do not revert. Let's see how this goes. I'm sure that after `make www` nothing will be changed.

**Comment by xexyl** [comment]
And so you know: the change in terms was just that: the functionality was fine as the commands were swapped. The only change here is the tool names.

**Comment by xexyl** [comment]
In fact reverting to that commit you suggest would remove the fixes to the issue Leo opened, too, plus I think the html.sed file.

**Comment by xexyl** [comment]
> > So please do not revert. Let's see how this goes. I'm sure that after `make www` nothing will be changed.
> 
> Can you make some PRs that fix things in the next day or two? If you can, we would be happy to not revert.

I can. As soon as make www goes fine I will. I just have to let it go through. If I don't get the commit in today I'll do so tomorrow morning.

**Comment by lcn2** [comment]
> In fact reverting to that commit you suggest would remove the fixes to the issue Leo opened, too, plus I think the html.sed file.

If we reverted, then certain editorial and typo and related fixes would be re-applied to the set.

**Comment by xexyl** [comment]
> > In fact reverting to that commit you suggest would remove the fixes to the issue Leo opened, too, plus I think the html.sed file.
> 
> If we reverted, then certain editorial and typo and related fixes would be re-applied to the set.

Much more would be lost too, though.

**Comment by xexyl** [comment]
... and there are probably more than you realise. I have no idea what I did with 2018/ferguson and 2020/ferguson2 for instance. Nu checker was fixed. Many other things. It would be a nightmare and it would solve nothing as it was just a terminology swap. The unicode symbols are already being shown correctly.

Keep in mind that although yes the tools were renamed (in error it seems) the tools were also changed here so there was no functional difference here; the only things changed were names. That's it.

**Comment by lcn2** [comment]
> > > In fact reverting to that commit you suggest would remove the fixes to the issue Leo opened, too, plus I think the html.sed file.
> > 
> > 
> > If we reverted, then certain editorial and typo and related fixes would be re-applied to the set.
> 
> Much more would be lost too, though.

We are faced with a difficult choice .. not hold IOCCC28 until sometime next year, or revert to the web site state before we went on vacation and them selectively applying the edits on top of that.

Right now it appears that things not in a stable shape.  But as you suggested that can be fixed. 

We are in an extreme travel complications that will take is a few days to get out of .. and we will he off the internet again for about 2 days.

Can you get this repo and the other repo into good shape by then?  If so we would be very happy to go with that!

**Comment by xexyl** [comment]
> > > > In fact reverting to that commit you suggest would remove the fixes to the issue Leo opened, too, plus I think the html.sed file.
> > > 
> > > 
> > > If we reverted, then certain editorial and typo and related fixes would be re-applied to the set.
> > 
> > 
> > Much more would be lost too, though.
> 
> We are faced with a difficult choice .. not hold IOCCC28 until sometime next year, or revert to the web site state before we went on vacation and them selectively applying the edits on top of that.
> 
> Right now it appears that things not in a stable shape. But as you suggested that can be fixed.
> 
> We are in an extreme travel complications that will take is a few days to get out of .. and we will he off the internet again for about 2 days.
> 
> Can you get this repo and the other repo into good shape by then? If so we would be very happy to go with that!

Like I said the html files are fine it's just the tool names. I already have in my clone the tools swapped again and just checking `make www` to make sure everything goes fine.

But right now the html files are as they should be: unicode symbols and so on. There should be no delay in the contest at all. 

So yes I can get this repo fine. Just waiting for `make www` to be done. As long as no html files changed (besides `bin/index.html` due to change in tool names) I cannot imagine any other html file changing. Just `bin/` files.

**Comment by xexyl** [comment]
.. and safe travels!

**Comment by xexyl** [comment]
`make www` reported no problems! After you merge it should be fine. I had a thought on the jstrdecode issue with `"`: it could be that if it's the first and last character it would be valid but otherwise not. I have this vague memory that there already exists this functionality so maybe a flag has to be passed; not sure but I have to go now.

**Comment by xexyl** [comment]
Ah .. encoding has the skip quotes. Maybe decoding needs that as well?

**Comment by lcn2** [comment]
> Ah .. encoding has the skip quotes. Maybe decoding needs that as well?

## IMPORTANT functionality

Be default, `jstrencode(1)` **MUST** append and prepend double quotes.  So:

```sh
$ jstrencode foo
"foo"
```

Be default, `jstrdecode(1)` **MUST** expect enclosing double quotes or generate an ERROR. So:

```sh
$ jstrdecode foo
jstrdecode ERROR: ...
```

There are, or should be, command line options flags that "deal with" enclosing double quotes.  

For example, a flag to ask `jstrencode(1)` to **NOT** enclose the output in double quotes. 

For example, a flag to ask `jstrdecode(1)` to **NOT** require enclosing double quotes, and thus **NOT** throw an error if the input/arg(s) lack them. 

There are, or were, or should have been such command line options flags .. we are not near a terminal to check .. so if such options exist, please be sure they still work.  If no such options exists (or no longer exist), then please add them. 

We'd check the man pages in the repo via a browser, but the current [https://github.com/ioccc-src/mkiocccentry/blob/master/jparse/man/man1/jstrencode.1](https://github.com/ioccc-src/mkiocccentry/blob/master/jparse/man/man1/jstrencode.1) contains text for `jstrdecode(1)`, so we cannot advise of what they had/have in terms command line options.

**Comment by lcn2** [comment]
# Recommend Priorities

We recommend a priority on moving to a state where we can do a "soft code freeze 🥶" for the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) while making sure that `make quick_www` and then `make www` work **AFTER** installation of the tools from [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) and the [jparse sub-directory of mkiocccentry](https://github.com/ioccc-src/mkiocccentry/tree/master/jparse) and **AFTER** import and JSON string encoding and the decoding of JSON encoded strings are working. 

In particular, fixing stuff (code, code comments,man pages, doc) that flipped around for `jstrencode(1)` and `jstrdecode(1)` while making sure that the tools, after they are installed, work well with `make quick_www` and then `make www`.

We are aware that there are questions and issues relating to blackslashes and perhaps Unicode/UTF-8 matters. While those are interesting and important, they may not necessarily be a problem for the existing web site and thus are not a top priority.  

Once [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) is stable and in a position to have a "soft code freeze 🥶", we can consider those "questions and issues" because by then, `make www` will be happy 😊, we can make progress on the **Great Fork Merge** 🥳.

And by consider those "questions and issues", assess the size of the task to investigate 🔬, determine the effort to answer open questions, assess the potential impact on the web site, and if there is enough time, install a code change into the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) before the "hard code freeze 🧊" and final release prior to IOCCC28. 

We hope this helps.

**Comment by xexyl** [comment]
Ah, it's this thread. Okay I'll try replying now but might take a bit of time.

**Comment by xexyl** [comment]
> # Man page bugs
> Perhaps the man pages have been fixed in some PR that we have not yet considered:
> 
> We'd check the man pages in the repo via a browser, but the current https://github.com/ioccc-src/mkiocccentry/blob/master/jparse/man/man1/jstrencode.1 contains text for `jstrdecode(1)`, so we cannot advise of what they had/have in terms command line options.

Thanks for pointing this out! I forgot to do the name swap when I was rushing yesterday. I did that a few minutes ago. Syncing to mkiocccentry now too.

**Comment by xexyl** [comment]
And no: jparse in mkiocccentry is not up to date because I made some slight changes yesterday. Nothing vital but something that I felt should be done. I'll sync it with the man pages as well. Still the jstrdecode issue with quotes will take more time to consider and it might need an option to say it's a quote. I'll explain in a bit.

**Comment by xexyl** [comment]
Ugh .. I just found yet another problem from the rush! I'll fix that too.

**Comment by xexyl** [comment]
Doing a sync ... will then reply to other comments if I can. Otherwise later on today I'm sure.

**Comment by xexyl** [comment]
> > Ah .. encoding has the skip quotes. Maybe decoding needs that as well?
> 
> ## IMPORTANT functionality
> Be default, `jstrencode(1)` **MUST** append and prepend double quotes. So:
> 
> ```shell
> $ jstrencode foo
> "foo"
> ```
> 
> Be default, `jstrdecode(1)` **MUST** expect enclosing double quotes or generate an ERROR. So:
> 
> ```shell
> $ jstrdecode foo
> jstrdecode ERROR: ...
> ```
> 

Unfortunately it's not that simple. Because this is also valid json:

```json
true
```

as is

```json
-5
```

and so on. If we expected quotes then it would be in error. Instead we need an option to jstrdecode that says the arg is a string and then if that arg is used we expect a `"` as the first and last character. I can do that I hope today maybe even in a bit if I can find the energy.

> There are, or should be, command line options flags that "deal with" enclosing double quotes.
> 

There are but not in the same way you refer to.

> For example, a flag to ask `jstrencode(1)` to **NOT** enclose the output in double quotes.
> 

Okay but in this case it should actually be that using it means TO enclose the output in double quotes. Why? Because otherwise we'll get something like:

```sh
$ jstrencode -- -5
"-5"
```

which is wrong.

I suggest the option `-s` to tell it that it is meant to be a string.

> For example, a flag to ask `jstrdecode(1)` to **NOT** require enclosing double quotes, and thus **NOT** throw an error if the input/arg(s) lack them.
> 

In this case it has to be the opposite for reasons I cited; it should be that it IS a string. Otherwise the new option you suggest would have to be used most the time which is annoying to the user. Here too I suggest `-s`.


> There are, or were, or should have been such command line options flags .. we are not near a terminal to check .. so if such options exist, please be sure they still work. If no such options exists (or no longer exist), then please add them.
> 

There were options that deal with quotes but not like this. As I see it the following options should be added to jstrencode:


```sh
-s		assume args are strings, enclosing output in quotes
```

and jstrdecode should have:

```sh
-s		assume arg is a string, expecting a leading and trailing "
```

I hope to work on these today but as I said I am extremely tired .. woke up too early :( I hope I get more energy sometime soon.




> # Man page bugs
> Perhaps the man pages have been fixed in some PR that we have not yet considered:
> 
> We'd check the man pages in the repo via a browser, but the current https://github.com/ioccc-src/mkiocccentry/blob/master/jparse/man/man1/jstrencode.1 contains text for `jstrdecode(1)`, so we cannot advise of what they had/have in terms command line options.


That was fixed in jparse and mkiocccentry (along with other problems discovered). Thanks!

**Comment by xexyl** [comment]
> # Recommend Priorities
> We recommend a priority on moving to a state where we can do a "soft code freeze 🥶" for the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) while making sure that `make quick_www` and then `make www` work **AFTER** installation of the tools from [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) and the [jparse sub-directory of mkiocccentry](https://github.com/ioccc-src/mkiocccentry/tree/master/jparse) and **AFTER** import and JSON string encoding and the decoding of JSON encoded strings are working.

Well the good thing is that after you merge the commit I made today this part should be fine. Also a mistake with the rush to get those important fixes in yesterday was repaired.

> 
> In particular, fixing stuff (code, code comments,man pages, doc) that flipped around for `jstrencode(1)` and `jstrdecode(1)` while making sure that the tools, after they are installed, work well with `make quick_www` and then `make www`.
> 

This should, as noted above, be fine, though I have not tested it with the jparse subdirectory of mkiocccentry. However given that it's now synced up (though not merged in the master branch) it should be fine.

> We are aware that there are questions and issues relating to blackslashes and perhaps Unicode/UTF-8 matters, and perhaps other jparse matters. While those are interesting and important, they may not necessarily be a problem for the existing web site and thus are not a top priority.

I agree there. I can always work on it in the interim of course.

> 
> Once [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) is stable and in a position to have a "soft code freeze 🥶", we can consider those "questions and issues" because by then, `make www` will be happy 😊, we can make progress on the **Great Fork Merge** 🥳.

Great. I hope that the new options I described will be easy enough. I don't know if you want those in the mkiocccentry repo or not. I can guess that you do want them in and I also guess you want the minimum version of the jstrdecode in the bin tools (here) updated too. I can do that for sure.

> 
> And by consider those "questions and issues", assess the size of the task to investigate 🔬, determine the effort to answer open questions, assess the potential impact on the web site, and if there is enough time, install a code change into the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) before the "hard code freeze 🧊" and final release prior to IOCCC28.

That makes sense.


> 
> ## UPDATE 0a
> ### Question
> Is the [mkiocccentry repo](https://github.com/ioccc-src/mkiocccentry) "in sync" with the [jparse repo](https://github.com/xexyl/jparse)? We ask this because we thought we say some slight differences in `jparse/json_util.c` via the browser but we cannot say for certain.

It is now thanks to your comment .. you were right: that file did have some changes as did some others. Some of them I am not surprised as I made those changes after the pull request. Others surprised me. I thought I had synced them. But obviously I didn't. The man page files were swapped too thanks to you. I guess I shouldn't have tried to rush it when I had to leave early morning but at least it should be fine now and we also found some other issues in the process.

**Comment by xexyl** [comment]
> 
> We hope this helps.

!sknahT .did yllautca ti ,yako ... ton did tI

**Comment by xexyl** [comment]
I guess that this issue here though can be closed as complete? Unless you want the new options to the encode/decode tools in first. I'm going to go afk a bit .. when back I hope I can work on the new options.

**Comment by xexyl** [comment]
The `jstrdecode(1)` new option `-s` has been, I believe, completed.

The jstrencode now needs its version.

**Comment by xexyl** [comment]
The jstrencode version was done a bit ago .. doing a make test. Increased version of each to 2.5 as it seems like a significant update.

Also you might enjoy the modification to the help string of the `-q` option to all the tools. I think I might do that for the mkiocccentry repo too or at least txzchk but we'll see.

After these are synced to mkiocccentry I'll do a make install from the mkiocccentry subdirectory and then do a make www though I can't imagine there will be any problems. I will also update the minimum version of the tool in the bin/ scripts.

**Comment by xexyl** [comment]
Well okay a test did fail .. I wonder why. Probably the new flag but I'll check.

**Comment by xexyl** [comment]
Hmm ... strange issue. Working on it.

**Comment by xexyl** [comment]
Oh. I think I see. Investigating.

**Comment by xexyl** [comment]
Solved. Will do a commit soonish .. then sync to mkiocccentry then work on the website scripts.

**Comment by xexyl** [comment]
Committed to the jparse repo. Will now make sure it goes well in test workflow. Assuming all is good I'll sync to mkiocccentry and then update the bin/ tools here. After that I'll run `make www` and make sure things are okay. I will then make a commit here if all is good. I don't think I'll be doing anything else today with code, though.

**Comment by lcn2** [comment]
> Unfortunately it's not that simple. Because this is also valid json:
> 
> ```json
> true
> ```
> 
> as is
> 
> ```json
> -5
> ```

The `jstrencode(1)` and `jstrdecode(1)` tools work on strings and JSON encoded strings.  They do **NOT** work on JSON booleans, JSON numbers, JSON nulls, etc.

So the above JSON examples to not apply.

**Comment by lcn2** [comment]
Thank you for resolving this imporant issue @xexyl

**Comment by xexyl** [comment]
> > Unfortunately it's not that simple. Because this is also valid json:
> > ```json
> > true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > as is
> > ```json
> > -5
> > ```
> 
> The `jstrencode(1)` and `jstrdecode(1)` tools work on strings and JSON encoded strings. They do **NOT** work on JSON booleans, JSON numbers, JSON nulls, etc.
> 
> So the above JSON examples to not apply.

Hmm .... that is a valid point indeed given the name. However the examples have always shown that and I think that also came from you, though perhaps not. But the name would suggest that it's only strings. On the one hand it would be nice to have a general encoder/decoder but on the other ...

I wonder what should be done. Probably need to rethink this and then maybe take the -s option away. It might be good to still have the option I was working on though, to validate as JSON or not. But it might not. I'll close the pull request over at the mkiocccentry repo for now.

**Comment by xexyl** [comment]
Or actually I'll turn it into a draft.

And you're welcome.

**Comment by xexyl** [comment]
My guess is the following.

Given the name I can get rid of the `-s` option.

The library change is still good, however.

Then I can also use the new options I added that have the parser validate it.

That being said if we do this then examples have to be heavily modified.

**Comment by lcn2** [comment]
Hello from eastern Greenland 🇬🇱 👋

**Comment by xexyl** [comment]
Actually .. it turns out that the use of the library update is very useful! It's just that we now will assume it's only strings.

I have added a `-j` and `-J level` option to `jstrdecode(1)`.

If `-j` is specified do not validate as JSON (the reason for doing it normally is it helps out with the decoding). Otherwise do validate it. If validation is desired (`-j` not used) then `-J` sets the JSON debug level, of course.

I have already removed the `-s` option. I have to do the same in `jstrencode` too. I'll then make a commit over there and then mkiocccentry and make it not a draft.

**Comment by xexyl** [comment]
> Hello from eastern Greenland 🇬🇱 👋

And yet .. sadly, due to human arrogance, it's getting more green - in colour, and less green in environment :(

Anyway glad travels are going well! Stay safe and well!

**Comment by xexyl** [comment]
Unsure if jstrdecode is supposed to quote the output now .. I think not since it's decoding, not encoding. I'll do it that way but if you can think of a case where an option would be good for that, please let me know.

**Comment by lcn2** [comment]
> Unsure if jstrdecode is supposed to quote the output now .. I think not since it's decoding, not encoding. I'll do it that way but if you can think of a case where an option would be good for that, please let me know.

By default, `jstrdecode(1)` does **NOT** enclose the output in double quotes.

If `-Q` or `-e` is given, then it will.

**Comment by xexyl** [comment]
> 



> > Unsure if jstrdecode is supposed to quote the output now .. I think not since it's decoding, not encoding. I'll do it that way but if you can think of a case where an option would be good for that, please let me know.
> 
> By default, `jstrdecode(1)` does **NOT** enclose the output in double quotes.
> 
> If `-Q` or `-e` is given, then it will.

Thanks for the reminder!
